### PR TITLE
pageserver: only apply `ClearVmBits` on relevant shards

### LINKED ISSUE #9855: pageserver: slow `get_rel_exists()` during WAL ingestion with many relations
In a customer workload, we saw slow pageserver WAL ingestion with CPU pinned at 100%. A [CPU profile](https://github.com/user-attachments/files/17872227/profile.svg.gz) revealed a significant amount of time spent deserializing relations for `get_relsize()`, in `get_rel_exists()`. The customer has ~100,000 relations, so this is likely quadratic behavior for large relation counts (see also #9516).

These relation sizes are cached. However, indications are that these calls are for relations that don't exist.

> I think we're getting `get_rel_exists() == false` in all of these cases. We see `Version::get()` under `get_rel_exists()` take 4% in the profile. If the relation did exist, we would drop into `get_rel_size()` which again calls `Version::get()` before updating the cache. Since we're only seeing `Version::get()` from `get_rel_exists`, not from `get_rel_size`, we're returning before calling `get_rel_size`, which only happens when the relation doesn't exist.
>
>Unless, of course, we're hitting a cache on the second `Version::get()` call.

@jcsp also says:

> I don't like that the ingest code even goes to Timeline to ask about relation sizes most of the time (thrashes a mutex): it should keep a running cache within the DatadirModification since the ingest is the only thing that modifies these sizes.

### PR DESCRIPTION
**Pull request by erikgrinaker**
# Problem

VM (visibility map) pages are stored and managed as any regular relation page, in the VM fork of the main relation. They are also sharded like other pages. Regular WAL writes to the VM pages (typically performed by vacuum) are routed to the correct shard as usual. However, VM pages are also updated via `ClearVmBits` metadata records emitted when main relation pages are updated. These metadata records were sent to all shards, like other metadata records. This had the following effects:

* On shards responsible for VM pages, the `ClearVmBits` applies as expected.

* On shard 0, which knows about the VM relation and its size but doesn't necessarily have any VM pages, the `ClearVmBits` writes may have been applied without also having applied the explicit WAL writes to VM pages.

* If VM pages are spread across multiple shards (unlikely with 256MB stripe size), all shards may have applied `ClearVmBits` if the pages fall within their local view of the relation size, even for pages they do not own.

* On other shards, this caused a relation size cache miss and a DbDir and RelDir lookup before dropping the `ClearVmBits`. With many relations, this could cause significant CPU overhead.

This is not believed to be a correctness problem, but this will be verified in #9914.

Resolves #9855.

# Changes

Route `ClearVmBits` metadata records only to the shards responsible for the VM pages.

Verification of the current VM handling and cleanup of incomplete VM pages on shard 0 (and potentially elsewhere) is left as follow-up work.

### CODE REVIEW

**Review by VladLazar — APPROVED** [review]
Looks fine in a pinch, but it's not clear to me why we would want to send these metadata records to the pageserver if they result in a no-op.

I would make the change in `wal_decoder::decoder::MetadataRecord::from_decoded` instead. More specifically, we should check if the keys in ClearVmBits belong to the current shard and skip the metadata record in that case.

**Comment by erikgrinaker** [comment]
> I would make the change in `wal_decoder::decoder::MetadataRecord::from_decoded` instead. More specifically, we should check if the keys in ClearVmBits belong to the current shard and skip the metadata record in that case.

Yep. We should consider doing that before merging, but I wanted a minimal, targeted fix that we could deploy to a test instance immediately without much vetting.

**Comment by erikgrinaker** [comment]
Split off some observability to https://github.com/neondatabase/neon/pull/9911 and rebased this on top.

**Comment by erikgrinaker** [comment]
I moved the filtering up into the decoder. I think this should be good to go, PTAL. Additional test coverage will be added as part of #9914.

File: libs/wal_decoder/src/decoder.rs [hunk 1]
```diff
@@ -89,19 +97,48 @@ impl MetadataRecord {
...
+                let vm_blk = pg_constants::HEAPBLK_TO_MAPBLOCK(heap_blk);
+                shard.is_key_local(&rel_block_to_key(clear_vm_bits.vm_rel, vm_blk))
+            };
+            // Send the old and new VM page updates to their respective shards.
+            clear_vm_bits.old_heap_blkno = clear_vm_bits
+                .old_heap_blkno
+                .filter(|&blkno| is_local_vm_page(blkno));
+            clear_vm_bits.new_heap_blkno = clear_vm_bits
+                .new_heap_blkno
+                .filter(|&blkno| is_local_vm_page(blkno));
```
Comment (VladLazar) [inline libs/wal_decoder/src/decoder.rs:134]: I was a bit surprised at this level of granularity, but looking at how the `old_heap_blkno` and `new_heap_blkno` are handled in `ingest_clear_vm_bits`, I think this is correct.

**Review by VladLazar — APPROVED** [review]

File: libs/wal_decoder/src/decoder.rs (hunk 1 above)
Comment (erikgrinaker) [reply libs/wal_decoder/src/decoder.rs:134]: Yeah, the blocks are processed individually (unless they're equal), and they each have to hit their respective shard if the VM relation straddles several shards.

**Pull request merged**